Media Related API Draft Review (#1617)
MicrosoftEdge/WebView2Feedback: specs/Media.md

WebView2 is a control that developers can implement for part of, or for the entirety of their app. In the either case, developers may need the ability to control the media within the webview's bounds. This API will allow developers to control the sound of each webview2 instance by enabling the mute/unmute of the media within the webview. There will also be the ability to check if the sound is muted or audible.

# Description
Enable end developer to `Mute`, `Unmute` a webview content. And ability to check for the `IsMuted` and `IsCurrentlyAudible`.

# Examples

The following code snippet demonstrates how the Media related API can be used:

## Win32 C++

```cpp
bool ViewComponent::HandleWindowMessage(
    HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam, LRESULT* result)
{
    if (message == WM_COMMAND)
    {
        case IDM_MUTE:
            m_webView->Mute();
            return true;
        case IDM_UNMUTE:
            m_webView->Unmute();
            return true;
        case IDM_IS_MUTED:
            BOOL isMuted;
            m_webView->get_IsMuted(&isMuted);
            MessageBox(nullptr, isMuted ? L"Yes" : L"No", L"Is Muted", MB_OK);
            return true;

>         case IDM_IS_CURRENTLY_AUDIBLE:

Reviewer: We'd probably want to demonstrate the events for the muted and audible properties. Like putting an icon at the end of the sample app's window's title when the values change like 🔇 for muted and 🔊 for audible.

Author: Make sense. I will make the change to code and also this spec

Author: updated a bit. Still debating on whether to use the IsMutedChanged event handler

    }

    void UnmuteCmdExecuted(object target, ExecutedRoutedEventArgs e)
    {
        webView.CoreWebView2.Unmute();
    }

    void IsMutedCmdExecuted(object target, ExecutedRoutedEventArgs e)
    {
        bool is_muted = webView.CoreWebView2.IsMuted;
        MessageBox.Show(this, is_muted? "Is muted": "Is Not muted", "Is Muted or Not");
    }

    void IsCurrentlyAudibleCmdExecuted(object target, ExecutedRoutedEventArgs e)
    {
        bool is_currently_audible = webView.CoreWebView2.IsCurrentlyAudible;
        MessageBox.Show(this, is_currently_audible ? "Is audible" : "Is not audible", "Is Currently Audible or Not");
    }
```

# API Notes

See [API Details](#api-details) section below for API reference.

# API Details

## Win32 C++

```IDL
interface ICoreWebView2StagingIsMutedChangedEventHandler;
interface ICoreWebView2StagingIsCurrentlyAudibleChangedEventHandler;

[uuid(76eceacb-0462-4d94-ac83-423a6793775e), object, pointer_default(unique)]
interface ICoreWebView2_2 : ICoreWebView2 {
  /// Adds an event handler for the `IsMutedChanged` event.
  /// `IsMutedChanged` runs when the mute state changes. The event may run 
  /// when `Mute` and `Unmute` are called.

>   HRESULT add_IsMutedChanged(

Author: I tried couple things like toggle the mute icon from the video tag. None of them change the mute/unmute state. But I would argue still keep the IsMutedChanged event, as user might need to listen to it to get the mute state change from using mute/unmute API. (like the sample code I wrote). Also I was wondering should we change the API naming to indicate that the mute and unmute is changing the mute state of the current window/tab (or just say it clearly in the documentation)

>   HRESULT add_IsMutedChanged(

Reviewer: If the end dev is calling Mute/Unmute then they can directly run the code they have in their event handler just after. This avoids having to setup an event handler and subscribe to the event making things simpler. Unless there's other ways for IsMuted to change other than Mute/Unmute

      [in] ICoreWebView2StagingIsMutedChangedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_IsMutedChanged`.
  HRESULT remove_IsMutedChanged(
      [in] EventRegistrationToken token);

  /// Mutes all audio output from this CoreWebView2.
  HRESULT Mute();
  /// Unmutes all audio output from this CoreWebView2.
  HRESULT Unmute();
  /// Indicates whether all audio output from this CoreWebView2 is muted or not.
  [propget] HRESULT IsMuted([out, retval] BOOL* isMuted);
  
  /// Adds an event handler for the `IsCurrentlyAudibleChanged` event.
  /// `IsCurrentlyAudibleChanged` runs when the audible state changes.
  HRESULT add_IsCurrentlyAudibleChanged(
      [in] ICoreWebView2StagingIsCurrentlyAudibleChangedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_IsCurrentlyAudibleChanged`.
  HRESULT remove_IsCurrentlyAudibleChanged(
      [in] EventRegistrationToken token);
  
  /// Indicates whether any audio output from this CoreWebView2 is audible.

>   /// CoreWebView2 can be audible when muted, IsCurrentlyAudible is used to indicate

Reviewer: I assume when you say 'can be audible when muted' you mean IsCurrentlyAudible can be true even when IsMuted is also true. Maybe a more explicit way to phrase this would be like the following:

```suggestion
  /// IsCurrentlyAudible is used to indicate if there is audio or media currently playing and can be true even if IsMuted is true or the volume is turned off.
```

>   /// if there are audio/media currently playing.

Reviewer: when you say 'audio/media' is that only media with audio? Like a video tag playing a video that has no audio (is that possible?)

Author: Looking at the comment below, it should only be any audio playing. I'm also trying to find a video with no audio to test that  

// Called when an audio change occurs to this WebContents. If |audible| is
  // true then one or more frames or child contents are emitting audio; if
  // false, then no frames or child contents are emitting audio. See
  // OnFrameAudioStateChanged for per-frame information.
virtual void OnAudioStateChanged(bool audible) {}

Author: I tried mute the video using the mute button within the video tag. If I mute the audio, the IsCurrentlyAudible value is false

Author: I added the IsCurrentlyAudible event hoping to help end developer detect whether content (media mainly) is playing, and they would be able to time out due to user inactive. But now thinking about it, IsCurrentlyAudible doesn't seems to be a good indication for that purpose. Also given the video tab behavior, the browser is treating IsCurrentlyAudible value as false when muted

Author: Best might be to remove this API for now and working on adding another API to indicate if content are playing so end developer can perform action due to user inactive. Would you think that would be better?

Reviewer: IsCurrentlyAudible seems like its still useful for showing the end user which WebView2 is making noise and letting them then mute that WebView2 like the speaker icon in the tab name / favicon of browser tabs.

>   [propget] HRESULT IsCurrentlyAudible([out, retval] BOOL* isAudible);

Reviewer: From your comment this isn't true if the CoreWebView2 is audible but is true if the CoreWebView2 is playing audio or media with audio. Should this be renamed `IsPlayingAudio`?

Author: Yeah it make sense to rename. I inherited the name from the chromium code, but it does sounds confusing

Reviewer: Or maybe IsAudioPlaying sounds better than IsPlayingAudio?

}

/// Implements the interface to receive `IsMutedChanged` events.  Use the
/// IsMuted method to get the mute state.
[uuid(B357DC3B-D4C3-4FDE-BF45-C11ECE606B98), object, pointer_default(unique)]
interface ICoreWebView2StagingIsMutedChangedEventHandler : IUnknown {
  /// Provides the event args for the corresponding event.  No event args exist
  /// and the `args` parameter is set to `null`.
  HRESULT Invoke([in] ICoreWebView2Staging2* sender, [in] IUnknown* args);
}

/// Implements the interface to receive `IsCurrentlyAudibleChanged` events.  Use the
/// IsCurrentlyAudible method to get the audible state.
[uuid(5DEF109A-2F4B-49FA-B7F6-11C39E513328), object, pointer_default(unique)]
interface ICoreWebView2StagingIsCurrentlyAudibleChangedEventHandler : IUnknown {
  /// Provides the event args for the corresponding event.  No event args exist
  /// and the `args` parameter is set to `null`.
  HRESULT Invoke([in] ICoreWebView2Staging2* sender, [in] IUnknown* args);
}
```

## .NET and WinRT

```c#

> namespace Microsoft.Web.WebView2.Core

Reviewer: To update this section you can run build-apiwriter and then copy paste the relevant sections out of the generated Microsoft.Web.WebView2.CoreWebView2.idl

Author: updated it

Reviewer: The events should show up in here too something like the following. If not let me know and we can figure out what's wrong

            event Windows.Foundation.TypedEventHandler<CoreWebView2, CoreWebView2ContextMenuRequestedEventArgs> ContextMenuRequested;

{
    public partial class CoreWebView2
    {

>         // There are other API in this interface that we are not showing 

Reviewer: Missing the events

Author: Seems like eventHandlers are not included in the Microsoft.Web.WebView2.CoreWebView2.idl. Should I still add it?

Reviewer: The events declaration should be included and the event args (if there are any for an event) but winrt & .net don't need to define the event handler interface

Author: There is no event args.

        public void Mute();

>         public void UnMute();

Reviewer: You've got the old UnMute name instead of Unmute:
```suggestion
        public void Unmute();
```

# Description
Enable end developer to `Mute`, `Unmute` a webview content. And ability to check for the `IsMuted` and `IsAudioPlaying`.

# Examples

The following code snippet demonstrates how the Media related API can be used:

## Win32 C++

```cpp
//! [IsAudioPlayingChanged] [IsAudioPlaying] [IsMuted] [Mute] [Unmute]
AudioComponent::AudioComponent(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
{
    // Register a handler for the IsAudioPlayingChanged event.
    // This handler just announces the audible state on the window's title bar.
    CHECK_FAILURE(m_webView->add_IsAudioPlayingChanged(
        Callback<ICoreWebView2StagingIsAudioPlayingChangedEventHandler>(
            [this](ICoreWebView2Staging2* sender, IUnknown* args) -> HRESULT {
                BOOL isAudioPlaying;
                CHECK_FAILURE(sender->get_IsAudioPlaying(&isAudioPlaying));

                BOOL isMuted;
                CHECK_FAILURE(sender->get_IsMuted(&isMuted));

                wil::unique_cotaskmem_string title;

>                 m_webView->get_DocumentTitle(&title);

Reviewer: Missing CHECK_ on get_DocumentTitle and on Mute/Unmute below

                    {
                        result = L"🔇 " + std::wstring(title.get());
                    }
                    else
                    {
                        result = L"🔊 " + std::wstring(title.get());
                    }
                }
                else
                {
                    result = std::wstring(title.get());
                }

                SetWindowText(m_appWindow->GetMainWindow(), result.c_str());
                return S_OK;
            })
            .Get(),
        &m_isAudioPlayingChangedToken));
}

// Mute the current window and show a mute icon on the title bar
 void AudioComponent::Mute()
 {
     m_webView->Mute();
 }

// Unmute the current window and hide the mute icon on the title bar
 void AudioComponent::Unmute()
 {
     m_webView->Unmute();
 }
 //! [IsAudioPlayingChanged] [IsAudioPlaying] [IsMuted] [Mute] [Unmute]
```

## .NET and WinRT

```c#
    void WebView_IsAudioPlayingChanged(object sender, object e)
    {
        bool isAudioPlaying = webView.CoreWebView2.IsAudioPlaying;
        bool isMuted = webView.CoreWebView2.IsMuted;
        string currentDocumentTitle = webView.CoreWebView2.DocumentTitle;

>         if (is_audio_playing)

Reviewer: ```suggestion
        if (isAudioPlaying)
```

            }
        }
        else
        {
            this.Title = currentDocumentTitle;
        }
    }
    
    void MuteCmdExecuted(object target, ExecutedRoutedEventArgs e)
    {
        webView.CoreWebView2.Mute();
    }

    void UnmuteCmdExecuted(object target, ExecutedRoutedEventArgs e)
    {
        webView.CoreWebView2.Unmute();
    }
```

# API Notes

See [API Details](#api-details) section below for API reference.

# API Details

## Win32 C++

```IDL
interface ICoreWebView2StagingIsAudioPlayingChangedEventHandler;

[uuid(71c906d9-4a4d-4dbe-aa1b-db64f4de594e), object, pointer_default(unique)]
interface ICoreWebView2_2 : ICoreWebView2 {
  /// Mutes all audio output from this CoreWebView2.
  ///
  /// \snippet AudioComponent.cpp Mute
  HRESULT Mute();

  /// Unmutes all audio output from this CoreWebView2.
  ///
  /// \snippet AudioComponent.cpp Unmute
  HRESULT Unmute();

  /// Indicates whether all audio output from this CoreWebView2 is muted or not.
  ///
  /// \snippet AudioComponent_Staging.cpp IsMuted
  [propget] HRESULT IsMuted([out, retval] BOOL* value);
  
  /// Adds an event handler for the `IsAudioPlayingChanged` event.

>   /// `IsAudioPlayingChanged` runs when the audible state changes.

Reviewer: ```suggestion
  /// `IsAudioPlayingChanged` is raised when the IsAudioPlaying property changes value.
```

  ///
  /// \snippet AudioComponent.cpp IsAudioPlayingChanged
  HRESULT add_IsAudioPlayingChanged(
      [in] ICoreWebView2StagingIsAudioPlayingChangedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

  /// Remove an event handler previously added with `add_IsAudioPlayingChanged`.
  HRESULT remove_IsAudioPlayingChanged(
      [in] EventRegistrationToken token);
  

>   /// Indicates whether any audio output from this CoreWebView2 is audible.

Reviewer: ```suggestion
  /// Indicates whether any audio output from this CoreWebView2 is playing.
```

>   /// CoreWebView2 can be audible when muted, IsAudioPlaying is used to indicate

Reviewer: ```suggestion
  /// This property will be true if audio is playing even if IsMuted is true.
```

  /// if there are audio currently playing.
  ///
  /// \snippet AudioComponent.cpp IsAudioPlaying
  [propget] HRESULT IsAudioPlaying([out, retval] BOOL* value);
}

/// Implements the interface to receive `IsAudioPlayingChanged` events.  Use the
/// IsAudioPlaying method to get the audible state.
[uuid(5DEF109A-2F4B-49FA-B7F6-11C39E513328), object, pointer_default(unique)]
interface ICoreWebView2StagingIsAudioPlayingChangedEventHandler : IUnknown {
  /// Provides the event args for the corresponding event.  No event args exist
  /// and the `args` parameter is set to `null`.

>   HRESULT Invoke([in] ICoreWebView2Staging2* sender, [in] IUnknown* args);

Reviewer: We've been doing ICoreWebView2* for event handler sender's args to make it possible to move to public without changing the interface

# Description
Enable end developer to `Mute`, `Unmute` a webview content. And ability to check for the `IsMuted` and `IsCurrentlyAudible`.

# Examples

The following code snippet demonstrates how the Media related API can be used:

## Win32 C++

```cpp
AudioComponent::AudioComponent(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
{
    //! [IsAudioPlayingChanged]
    // Register a handler for the IsCurrentlyAudibleChanged event.
    // This handler just announces the audible state on the window's title bar.
    CHECK_FAILURE(m_webView->add_IsAudioPlayingChanged(
        Callback<ICoreWebView2StagingIsAudioPlayingChangedEventHandler>(
            [this](ICoreWebView2Staging2* sender, IUnknown* args) -> HRESULT {
                BOOL isAudioPlaying;
                CHECK_FAILURE(sender->get_IsAudioPlaying(&isAudioPlaying));

                BOOL isMuted;
                CHECK_FAILURE(sender->get_IsMuted(&isMuted));

                wil::unique_cotaskmem_string title;
                m_webView->get_DocumentTitle(&title);
                std::wstring result = L"";

>                 //! [IsAudioPlaying]

Author: Do I put them all together like this ? `//! [IsAudioPlayingChanged] [IsMuted] [IsAudioPlaying]`

>                 //! [IsAudioPlaying]

Author: I moved it to upper level

# Background
WebView2 is a control that developers can implement for part of, or for the entirety of their app. In the either case, developers may need the ability to control the media within the webview's bounds. This API will allow developers to control the sound of each webview2 instance by enabling the mute/unmute of the media within the webview. There will also be the ability to check if the sound is muted or audible.

# Description
Enable end developer to `Mute`, `Unmute` a webview content. And ability to check for the `IsMuted` and `IsCurrentlyAudible`.

# Examples

The following code snippet demonstrates how the Media related API can be used:

## Win32 C++

```cpp
AudioComponent::AudioComponent(AppWindow* appWindow)
    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
{
    //! [IsCurrentlyAudibleChanged]
    // Register a handler for the IsCurrentlyAudibleChanged event.
    // This handler just announces the audible state on the window's title bar.
    CHECK_FAILURE(m_webView->add_IsCurrentlyAudibleChanged(
        Callback<ICoreWebView2StagingIsCurrentlyAudibleChangedEventHandler>(
            [this](ICoreWebView2Staging2* sender, IUnknown* args) -> HRESULT {

>                 BOOL isAudible;

Reviewer: These two event handlers are copies of one another. Probably want to move out to one method that you call from both.

    CHECK_FAILURE(m_webView->add_IsMutedChanged(
        Callback<ICoreWebView2StagingIsMutedChangedEventHandler>(
            [this](ICoreWebView2Staging2* sender, IUnknown* args) -> HRESULT {
                BOOL isAudible;
                CHECK_FAILURE(sender->get_IsCurrentlyAudible(&isAudible));
                
                BOOL isMuted;
                CHECK_FAILURE(sender->get_IsMuted(&isMuted));

                wil::unique_cotaskmem_string title;
                m_webView->get_DocumentTitle(&title);
                std::wstring result = (isAudible ? L"🔊 " : L"") + std::wstring(title.get()) + 
                    (isMuted ? L" 🔇" : L"");

                SetWindowText(m_appWindow->GetMainWindow(), result.c_str());
                return S_OK;
            })
            .Get(),
        &m_isMutedChangedToken));
    //! [IsMutedChanged]
}

//! [Mute]
// Mute the current window and show a mute icon on the title bar
 void AudioComponent::Mute()
 {
     m_webView->Mute();
 }
//! [Mute]

//! [Unmute]
// Unmute the current window and hide the mute icon on the title bar
 void AudioComponent::Unmute()
 {
     m_webView->Unmute();
 }
 //! [Unmute]
```

## .NET and WinRT

```c#

>     void MuteCmdExecuted(object target, ExecutedRoutedEventArgs e)

Reviewer: The .NET sample code should be doing the same thing as the C++ sample code

Author: updated
